src/serde.rs

```diff
@@ -1,20 +1,76 @@
 //! Multihash Serde (de)serialization
 
-use std::fmt;
+use core::{fmt, ptr, slice};
 
 use serde::{
     de::{self, SeqAccess, Visitor},
-    Deserialize, Deserializer, Serialize, Serializer,
+    ser, Deserialize, Deserializer, Serialize, Serializer,
 };
 
 use crate::Multihash;
 
+/// The maximum serialization size of `code` is 9 bytes (a large varint encoded u64) and for `size`
+/// is 2 bytes  (a large varint encoded u8), this makes a total of 11 bytes.
+const MAXIMUM_PREFIX_SIZE: usize = 11;
+
+/// The is currently no way to allocate an array that is some constant size bigger then a given
+/// const generic. Once `generic_const_exprs` are a thing, this struct will no longer be needed.
+/// Until then we introduce a hack. We allocate a struct, which contains two independent arrays,
+/// which can be specified with const generics. We then treat the whole struct as a slice of
+/// continuous memory.
+#[repr(C, packed)]
+struct Buffer<const SIZE_FIRST: usize, const SIZE_SECOND: usize> {
+    first: [u8; SIZE_FIRST],
+    second: [u8; SIZE_SECOND],
+}
+
+#[allow(unsafe_code)]
+impl<const SIZE_FIRST: usize, const SIZE_SECOND: usize> Buffer<SIZE_FIRST, SIZE_SECOND> {
+    fn new() -> Self {
+        let buffer = Self {
+            first: [0; SIZE_FIRST],
+            second: [0; SIZE_SECOND],
+        };
+
+        // Make sure that the struct allocated continuous memory, as we exploit that fact with the
+        // `as_slice` and `as_mut_slice()` methods.
+        let start_first = ptr::addr_of!(buffer.first) as *const u8;
+        let start_second = ptr::addr_of!(buffer.second) as *const u8;
+        unsafe {
+            // This should never happen, hence it's OK to panic.
+            if start_second.offset_from(start_first) != SIZE_FIRST as isize {
+                panic!("alignment of the struct is wrong")
+            }
+        }
+
+        buffer
+    }
+
+    fn len(&self) -> usize {
+        SIZE_FIRST + SIZE_SECOND
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        let start = ptr::addr_of!(self.first) as *const u8;
+        unsafe { slice::from_raw_parts(start, self.len()) }
+    }
+
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        let start = ptr::addr_of_mut!(self.first) as *mut u8;
+        unsafe { slice::from_raw_parts_mut(start, self.len()) }
```

Comment:
You should be able to write this as:

```suggestion
        unsafe { slice::from_raw_parts_mut(self as *mut u8, self.len()) }
```

Reply:
Nope:

```
error[E0606]: casting `&mut Buffer<SIZE_FIRST, SIZE_SECOND>` as `*mut u8` is invalid
  --> src/serde.rs:61:44
   |
61 |         unsafe { slice::from_raw_parts_mut(self as *mut u8, self.len()) }
   |                                            ^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0606`.
error: could not compile `multihash` (lib) due to previous error
```

Reply:
Well, `self as *mut Buffer<_, _> as *mut u8`.

Reply:
I went for `self as *mut _ as _`.